src/services/fields.ts

```diff
@@ -18,79 +25,88 @@ export type DetectedLabelsResponse = {
   detectedLabels: DetectedLabel[];
 };
 
-type ExtractedFieldsType = 'logfmt' | 'json' | 'mixed';
+export type DetectedField = {
+  label: string;
+  cardinality: number;
+  type: string;
+  parsers: string[];
+};
 
-interface ExtractedFields {
-  type: ExtractedFieldsType;
-  fields: string[];
-}
+export type DetectedFieldsResponse = {
+  fields: DetectedField[];
+};
 
-export function updateParserFromDataFrame(frame: DataFrame, sceneRef: SceneObject): ExtractedFields {
-  const variable = getLogsFormatVariable(sceneRef);
-  const res = extractParserAndFieldsFromDataFrame(frame);
-
-  let newType;
-  if (!res.type) {
-    newType = '';
-  } else if (res.type === 'mixed') {
-    newType = `| json  | logfmt | drop __error__, __error_details__`;
-  } else {
-    newType = ` | ${res.type}`;
+const getReducerId = memoize((sortBy: SortBy) => {
+  if (sortBy) {
+    const values: string[] = Object.values(ReducerID);
+    if (values.includes(sortBy)) {
+      return sortBy;
+    }
   }
+  return undefined;
+});
 
-  if (variable.getValue() !== newType) {
-    variable.changeValueTo(newType);
+export type ExtractedFieldsType = 'logfmt' | 'json' | 'mixed' | '';
+
+export function extractParserFromDetectedFieldParserFieldValue(parserString: string): ExtractedFieldsType {
+  switch (parserString) {
+    case 'json':
+      return 'json';
+    case 'logfmt':
+      return 'logfmt';
+    case '': // Structured metadata is empty
+      return '';
+    default: // if we get a parser with multiple
+      return 'mixed';
   }
-
-  return res;
 }
 
-export function extractParserAndFieldsFromDataFrame(data: DataFrame) {
-  const result: ExtractedFields = { type: 'logfmt', fields: [] };
-  const labelTypesField = data.fields.find((f) => f.name === 'labelTypes');
-  result.fields = Object.keys(
-    labelTypesField?.values.reduce((acc: Record<string, boolean>, value: Record<string, string>) => {
-      Object.entries(value)
-        .filter(([_, v]) => v === 'P')
-        .forEach(([k]) => (acc[k] = true));
-      return acc;
-    }, {}) ?? {}
-  );
-
-  const types: ExtractedFieldsType[] = [];
-  const linesField = data.fields.find((f) => f.name === 'Line' || f.name === 'body');
-
-  if (!linesField) {
-    return result;
+export function extractParserFieldFromParserArray(parsers?: string[]) {
+  const parsersSet = new Set(parsers?.map((v) => v.toString()) ?? []);
+
+  // Structured metadata doesn't change the parser we use, so remove it
+  parsersSet.delete('');
```

Comment:
If we delete `''`, do we need a `case '':` in `extractParserFromDetectedFieldParserFieldValue`? If not, then extractParserFromDetectedFieldParserFieldValue is not needed.

Reply:
extractParserFromDetectedFieldParserFieldValue, now `extractParserFromString` is called in more places then here, you are correct that case statement isn't needed for this calling context, but `extractParserFromString` is still needed as we could have multiple parsers for each field, i.e. the set value could be "logfmt, json".

This is confusing though, i wonder if we used something instead of an empty string for structured metadata if that would make it easier to read?

Reply:
@matyax this has been updated, should be a bit better now?

Reply:
Much better! Thank you.